pkg/tenancy/write/write.go

```diff
@@ -0,0 +1,72 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package write
+
+import (
+	"fmt"
+
+	"github.com/timescale/promscale/pkg/prompb"
+	"github.com/timescale/promscale/pkg/tenancy/config"
+)
+
+// writeAuthorizer is a write authorizer that authorizes if the incoming write request is valid to be written or not.
+type writeAuthorizer struct {
+	config.Config
+}
+
+var errTenantMismatch = fmt.Errorf("__tenant__ value and tenant-name from headers are different")
+
+// NewAuthorizer returns a new plainWriteAuthorizer.
+func NewAuthorizer(config config.Config) Authorizer {
+	return &writeAuthorizer{config}
+}
+
+func (a *writeAuthorizer) IsAuthorized(tenantName string) error {
+	if a.IsTenantAllowed(tenantName) {
+		return nil
+	}
+	return fmt.Errorf("%s %s", ErrUnauthorizedTenant.Error(), tenantName)
+}
+
+func (a *writeAuthorizer) VerifyAndApplyTenantLabel(tenantNameFromHeader string, labels []prompb.Label) ([]prompb.Label, error) {
```

Comment:
This seems really complex. Can we simplify by splitting into a few functions:

```
func (a *writeAuthorizer) getTenantLabelMatchingHeader(tenantNameFromHeader string, labels []prompb.Label) ([]prompb.Label, error) {
	for _, label := range labels {
		if label.Name == config.TenantLabelKey {
			if label.Value == tenantNameFromHeader {
				return labels, nil
			}
			if label.Value = "" {
				// Tenant label exists but no tenant value. This is invalid.
				return labels, fmt.Errorf("%s exists with an empty value", config.TenantLabelKey)
			}
			return labels, errTenantMismatch
		}
	}
	labels = append(labels, prompb.Label{Name: config.TenantLabelKey, Value: tenantNameFromHeader})
	return labels, nil
}



func (a *writeAuthorizer) getTenantNameFromLabel(labels []prompb.Label) (string) {
	for _, label := range labels {
		if label.Name == config.TenantLabelKey {
			//any value including empty is valid 
			return labels.Value
		}
	}
	return  ""
}

func (a *writeAuthorizer) VerifyAndApplyTenantLabel(tenantNameFromHeader string, labels []prompb.Label) ([]prompb.Label, error) {
	if tenantNameFromHeader != "" {
		return getTenantLabelMatchingHeader(tenantName, labels)
	}

	tenantName := getTenantNameFromLabel(labels)	
	if tenantName == "" {
		//can we integrate this inside of IsAuthorized if given the empty string?
		if a.AllowNonTenants() {
			// Non-multi-tenant write request.
			// We accept non-multi-tenant write requests in tenancy mode.
			return labels, nil
		} else {
			return labels, fmt.Errorf("non multi-tenant writes not allowed")
		}
	}
	return labels, a.IsAuthorized(tenantName)
}
```

Reply:
The code as-is has high Cyclomatic complexity making it hard to verify

Reply:
Simplified with another but similar function splitting.